algebird-core/src/main/scala/com/twitter/algebird/Scan.scala

```diff
@@ -0,0 +1,361 @@
+    *         where we don't pollute the {{State}} by pairing it redundantly with {{Unit}}.
+    */
+  def joinWithInput[I1 <: I]: Aux[I1, State, (I1, O)] = new Scan[I1, (I1, O)] {
+    override type State = self.State
+
+    override def initialState: State = self.initialState
+
+    override def presentAndNextState(i: I1, stateBeforeProcessingI: State): ((I1, O), State) = {
+      val (o, stateAfterProcessingA) = self.presentAndNextState(i, stateBeforeProcessingI)
+      ((i, o), stateAfterProcessingA)
+    }
+  }
+
+  def joinWithPriorState: Aux[I, State, (State, O)] = new Scan[I, (State, O)] {
+    override type State = self.State
+
+    override def initialState: State = self.initialState
+
+    override def presentAndNextState(i: I, stateBeforeProcessingI: State): ((State, O), State) = {
+      val (o, stateAfterProcessingA) = self.presentAndNextState(i, stateBeforeProcessingI)
+      ((stateBeforeProcessingI, o), stateAfterProcessingA)
+    }
+  }
+
+  def joinWithPosteriorState: Aux[I, State, (O, State)] = new Scan[I, (O, State)] {
+    override type State = self.State
+
+    override def initialState: State = self.initialState
+
+    override def presentAndNextState(i: I, stateBeforeProcessingI: State): ((O, State), State) = {
+      val (c, stateAfterProcessingA) = self.presentAndNextState(i, stateBeforeProcessingI)
+      ((c, stateAfterProcessingA), stateAfterProcessingA)
+    }
+  }
+
+  def joinWithIndex: Aux[I, (State, Long), (O, Long)] = join(Scan.index)
+
+  def zip[I2, O2](scan2: Scan[I2, O2]): Aux[(I, I2), (State, scan2.State), (O, O2)] = new Scan[(I, I2), (O, O2)] {
+    override type State = (self.State, scan2.State)
+
+    override def initialState: State = (self.initialState, scan2.initialState)
+
+    override def presentAndNextState(
+        i1i2: (I, I2),
+        stateBeforeProcessingI1I2: State
+    ): ((O, O2), State) = {
+      val (o1, state1AfterProcesingI1) =
+        self.presentAndNextState(i1i2._1, stateBeforeProcessingI1I2._1)
+      val (o2, state2AfterProcesingI2) =
+        scan2.presentAndNextState(i1i2._2, stateBeforeProcessingI1I2._2)
+      ((o1, o2), (state1AfterProcesingI1, state2AfterProcesingI2))
+
+    }
+  }
+
+  def join[I2 <: I, O2](scan2: Scan[I2, O2]): Aux[I2, (State, scan2.State), (O, O2)] = new Scan[I2, (O, O2)] {
+    override type State = (self.State, scan2.State)
+
+    override def initialState: State = (self.initialState, scan2.initialState)
+
+    override def presentAndNextState(i: I2, stateBeforeProcessingI: State): ((O, O2), State) = {
+      val (o1, state1AfterProcesingI1) = self.presentAndNextState(i, stateBeforeProcessingI._1)
+      val (o2, state2AfterProcesingI2) = scan2.presentAndNextState(i, stateBeforeProcessingI._2)
+      ((o1, o2), (state1AfterProcesingI1, state2AfterProcesingI2))
+    }
+  }
+
+  def compose[P](scan2: Scan[O, P]): Aux[I, (State, scan2.State), P] = new Scan[I, P] {
```

Comment:
maybe add comments to the methods: `this feeds each output of this into scan2` and join is like `this forks each input and feeds it into both this and scan2`, etc...

Reply:
I've added a bunch of comments. Let me know if you like the wording/framing of them.